src/dac.rs

```diff
@@ -0,0 +1,232 @@
+//Based on stm32hal by David-OConnor
+
+use cortex_m::asm;
+
+use crate::{
+    gpio::{self, Analog},
+    pac::{
+        DAC1,
+        dac1,
+    },
+    rcc::{Clocks, AHB},
+};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+
+use crate::pac::DMA1::{self};
+
+use crate::{
+    pac::{
+        self,
+        rcc,
+        RCC,
+    },
+};
+
+
+pub enum DacDevice {
+    One,
+    Two,
+}
+
+pub enum DacChannel {
+    One,
+    Two,
+}
+
+pub enum DacBitAlignment {
+    /// Eight bit precision, right-aligned.
+    EightRight,
+    /// 12-bit precision, left-aligned.
+    TwelveLeft,
+    /// 12-bit precision, right-aligned.
+    TwelveRight,
+}
+
+pub enum Trigger {
+    /// Timer 6
+    Tim6 = 0b000,
+    /// Timers 3 or 8
+    Tim3_8 = 0b001,
+    /// Timer 7
+    Tim7 = 0b010,
+    /// Timer 15
+    Tim5 = 0b011,
+    /// Timer 2
+    Tim2 = 0b100,
+    /// Timer 4
+    Tim4 = 0b101,
+    /// Eg, for interrupts
+    Exti9 = 0b110,
+    /// A software trigger
+    Swtrig = 0b111,
+}
+
+/// Mask selector for noise generation or Amplitude selector for triangle generation
+pub enum Mamp {
+     
+    /// Unmask bit0 of LFSR/ triangle amplitude equal to 1
+    Mamp0 = 0b0000,
+    /// Unmask bits[1:0] of LFSR/ triangle amplitude equal to 3
+    Mamp1 = 0b0001,
+    /// Unmask bits[2:0] of LFSR/ triangle amplitude equal to 7
+    Mamp2 = 0b0010,
+    /// Unmask bits[3:0] of LFSR/ triangle amplitude equal to 15
+    Mamp3 = 0b0011,
+    /// Unmask bits[4:0] of LFSR/ triangle amplitude equal to 31
+    Mamp4 = 0b0100,
+    /// Unmask bits[5:0] of LFSR/ triangle amplitude equal to 63 
+    Mamp5 = 0b0101, 
+    /// Unmask bits[6:0] of LFSR/ triangle amplitude equal to 127 
+    Mamp6 = 0b0110,
+    /// Unmask bits[7:0] of LFSR/ triangle amplitude equal to 255
+    Mamp7 = 0b0111,
+    /// Unmask bits[8:0] of LFSR/ triangle amplitude equal to 511 
+    Mamp8 = 0b1000,
+    /// Unmask bits[9:0] of LFSR/ triangle amplitude equal to 1023 
+    Mamp9 = 0b1001,
+    /// Unmask bits[10:0] of LFSR/ triangle amplitude equal to 2047
+    Mamp10 = 0b1010, 
+    // Unmask bits[11:0] of LFSR/ triangle amplitude equal to 4095
+    Mamp11 = 0b1011, 
+}
+
+/// Represents a Digital to Analog Converter (DAC) peripheral.
+pub struct Dac {
+    pub regs: DAC1,
+    device: DacDevice,
+    bits: DacBitAlignment,
+    vref: f32,
+}
+
+// todo: Calculate the VDDA vref, as you do with onboard ADCs!
+impl Dac 
+{
+    /// Initialize a DAC peripheral, including  enabling and resetting
+    /// its RCC peripheral clock. `vref` is in volts.
+    pub fn new(regs: DAC1, device: DacDevice, bits: DacBitAlignment, vref: f32) -> Self {
+        
+            let rcc = unsafe { &(*RCC::ptr()) };
+            match device {
+                DacDevice::One => {
```

Comment:
This method is always passed `DAC1`, doesn't that make this (and `DacDevice`) unnecessary?